File: backend/app/models/notification.py
import uuid
from datetime import datetime
from enum import StrEnum

from sqlalchemy import DateTime, ForeignKey, String, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base, pg_enum


class NotificationKind(StrEnum):
    OUTBID = "outbid"
    AUCTION_WON = "auction_won"
    AUCTION_LOST = "auction_lost"
    PROPERTY_APPROVED = "property_approved"
    PROPERTY_REJECTED = "property_rejected"
    KYC_REVIEWED = "kyc_reviewed"


class Notification(Base):
    __tablename__ = "notifications"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("users.id", ondelete="CASCADE"), index=True
    )
    kind: Mapped[NotificationKind] = mapped_column(pg_enum(NotificationKind, "notification_kind"))
    message: Mapped[str] = mapped_column(String(300))
    # What the notification is about, so the client can link straight to it. Both may be null.
    auction_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("auctions.id", ondelete="CASCADE"), default=None
    )
    property_id: Mapped[uuid.UUID | None] = mapped_column(
        ForeignKey("properties.id", ondelete="CASCADE"), default=None
    )
    read_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), default=None)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

File: frontend/types/notification.ts
export type NotificationKind =
  | "outbid"
  | "auction_won"
  | "auction_lost"
  | "property_approved"
  | "property_rejected"
  | "kyc_reviewed"
  | "new_message"
  | "support_ticket";

export interface Notification {
  id: string;
  kind: NotificationKind;
  message: string;
  auction_id: string | null;
  property_id: string | null;
  read_at: string | null;
  created_at: string;
}

export interface NotificationPage {
  items: Notification[];
  unread: number;
}

export interface MarkReadRequest {
  ids?: string[];
}
